### ISSUE TITLE
Dummy question about reportUnsupportedDunderAll / __all__ again 

### ISSUE THREAD

**Issue by expend20** [microsoft/pyright#3304]
**Describe the bug**
Static assigning `__all__` for the subdirectory is not recognized?

E.g.:
```python
__all__ = ['adb'] # adb is a subdirectory
```

**To Reproduce**

Just trying to use `pwntools` 
![image](https://user-images.githubusercontent.com/36543551/162060944-2fd5b9da-8c50-48ca-948b-c2f1c064a47c.png)
and at that level autocompletion is effectively stopped:
![image](https://user-images.githubusercontent.com/36543551/162061241-8d498c99-7de4-4d2c-9a91-c766f9b33c7e.png)
perhaps I should configure paths somehow?

**Expected behavior**

Being able to autocomple `pwn.pwnlib.util...` after `import pwn`

**Comment by erictraut** [comment]
What is the intent of adding `'adb'` to the `__all__` list? The `__all__` symbol indicates to the Python interpreter which symbols should be included when someone performs a wildcard import (`from x import *`) for the module. Is the intent for a wildcard import to include a symbol called `adb`?

The title of the issue references the `reportUnsupportedDunderAll` check, but the description of the issue doesn't reference this check. Is there a connection?

**Comment by expend20** [comment]
Hi @erictraut, thanks for quick reply! And sorry for choppy question.

> The title of the issue references the reportUnsupportedDunderAll check, but the description of the issue doesn't reference this check. Is there a connection?

Yep, it was on the first screenshot, when I'm trying to look at the error regarding `util` which is at the same level as `adb`. Let me show you once again, just to confirm it's working during python run:

![image](https://user-images.githubusercontent.com/36543551/162129633-2681ec8f-0173-4bef-9681-fc4e209d445b.png)

Let me summarize a bit:
* `__all__` is defined in `pwnlib/__init__.py` and there are subdirectories in the same folder `adb`, `utils`, etc
* When I open `pwnlib/__init__.py` I see errors like this: `[Pyright reportUnsupportedDunderAll] [W] "adb" is specified in __all__ but is not present in module` for each entry in the list of `__all__` 

> The __all__ symbol indicates to the Python interpreter which symbols should be included when someone performs a wildcard import (from x import *) for the module.

I don't know how actually how `__all__` should work, but let consider two scenarios with wildcard:

![image](https://user-images.githubusercontent.com/36543551/162131703-5480b5f6-1ed4-48f2-9493-91d843c960fb.png)

and without:

![image](https://user-images.githubusercontent.com/36543551/162131737-4d739753-e743-472c-b6a4-6bdcccb29061.png)
In both cases pycharm was able to complete the symbol.

**Comment by erictraut** [comment]
This module is doing something very unusual and suspect. It's relying on the presence of submodule names within an `__all__` list to implicitly import these submodules and populate them within the module's global namespace. All other libraries that I've seen have included explicit import statements such as `import adb` or `from . import adb`. Libraries should not rely on implicit (and as far as I know, undefined) side effects like this.

Supporting the pattern used here is difficult given the architecture of pyright. I'll think a bit more about how we might support this, but we may just close this issue as "not supported".

Perhaps you can convince the maintainers of this library to add explicit import statements so they don't rely on this implicit behavior.

In the meantime, the workaround is to import from `pwnlib` rather than `pwn` — as in `import pwnlib.adb` or `from pwnlib.adb import ADB`.

**Comment by AlexWaygood** [comment]
> This module is doing something very unusual and suspect. It's relying on the presence of submodule names within an `__all__` list to implicitly import these submodules and populate them within the module's global namespace. All other libraries that I've seen have included explicit import statements such as `import adb` or `from . import adb`.

In the stdlib, there are three modules that use this baffling implicit-`__all__`-import behaviour (they [cause problems](https://github.com/python/typeshed/issues/6523) for us over at typeshed):

- `email.__init__`
- `distutils.command.__init__`
- `xml.__init__`

I think using this "feature" of the Python import system perhaps used to be more common than it now is. I agree that if I were designing the Python import system from scratch today, I wouldn't include this "feature" in my design; it seems a bit of an antipattern to me. I don't think type checkers should need to support this behaviour.

**Comment by erictraut** [comment]
I did a bit more digging. As I mentioned above, the `pwnlib/__init__.py` module is including symbols in its `__all__` list that are not otherwise imported. I suspected that perhaps inclusion in the `__all__` list implicitly imported these submodules and included them within the top-level module's namespace. Upon further investigation, I've confirmed this isn't the case. You can see this from the Python repl by typing:
```
>>> import pwnlib
>>> print(dir(pwnlib))
['__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '__version__', 'absolute_import', 'args', 'commandline', 'config', 'context', 'device', 'term', 'timeout', 'version']
```

Note that symbols `adb`, `args`, etc. are not in the module's namespace despite appearing in the `__all__`. It is only after I execute `from pwnlib import *` that they appear.

So I think pyright's behavior is correct here. These symbols should not be included in the `pwnlib` module's namespace because they have not been explicitly imported by the module. This is a bug in the library, and you may want to report it to the maintainers.

I'm going to close this bug "as designed".

**Comment by expend20** [comment]
Thank you so much for investigating this!

**Comment by waweber** [comment]
@erictraut I think this issue should be reopened as there are some unaddressed nuances of `__all__`.

Refer to [the docs for `import`](https://docs.python.org/3.11/reference/simple_stmts.html?highlight=__all__#the-import-statement). It defines how `__all__` works for modules:

> The public names defined by a module are determined by checking the module’s namespace for a variable named `__all__`; if defined, it must be a sequence of strings which are names defined or imported by that module.

This lines up with everything so far. But there is additional behavior for packages which is detailed [in an entirely different section of the documentation for some reason](https://docs.python.org/3.11/tutorial/modules.html#importing-from-a-package):

>The import statement uses the following convention: if a package’s `__init__.py` code defines a list named `__all__`, it is taken to be the list of module names that should be imported when from package import * is encountered.

An example for a package `sound/effects/__init__.py` is given:

> ```python
> __all__ = ["echo", "surround", "reverse"]
> ```
> 
> This would mean that from sound.effects import * would import the three named submodules of the sound.effects package.

It also goes on to clarify that `__all__` can contain both names defined in `__init__.py` and names of submodules:

> 
> Be aware that submodules might become shadowed by locally defined names. For example, if you added a reverse function to the sound/effects/__init__.py file, the from sound.effects import * would only import the two submodules echo and surround, but not the reverse submodule, because it is shadowed by the locally defined reverse function:
> 
> ```python
> __all__ = [
>     "echo",      # refers to the 'echo.py' file
>     "surround",  # refers to the 'surround.py' file
>     "reverse",   # !!! refers to the 'reverse' function now !!!
> ]
> 
> def reverse(msg: str):  # <-- this name shadows the 'reverse.py' submodule
>     return msg[::-1]    #     in the case of a 'from sound.effects import *'
> ```
> 

### So to summarize:

- `__all__` should contain a list of names defined in the module/package
- The names are what is imported by a `*` import
- The list defines the package/module's public API
- If the file is a module:
  - `__all__` should contain a list of names defined in the module.
- If the file is a package's `__init__.py`:
  - If a name in `__all__` refers to a name defined in the `__init__.py`, that object is what will be imported, both by `from package import *` and `from package import name`.
  - If the name is not defined in `__init__.py`, it should refer to a submodule with that name, and that submodule will be imported by `from package import *` and `from package import name`

I think the rule needs to be enhanced to check for submodule names, as this appears to be correct behavior according to the docs, and as seen in the standard library packages mentioned by @AlexWaygood.

**Comment by erictraut** [comment]
@waweber, you're responding to a bug report that has been closed for well over a year. If you would like to report a bug, please open a new issue and fill out the bug template. In particular, please provide a specific example of a behavior that you think is incorrect and the behavior that you expect to see. From your post above, it's not clear to me what behavior you're seeing in pyright that you think is incorrect.
